providers/implementations/kdfs/hkdf.c

```diff
@@ -307,13 +307,11 @@ static int hkdf_common_set_ctx_params(KDF_HKDF *ctx, const OSSL_PARAM params[])
     }
 
     if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_SALT)) != NULL) {
-        if (p->data_size != 0 && p->data != NULL) {
-            OPENSSL_free(ctx->salt);
-            ctx->salt = NULL;
-            if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->salt, 0,
-                                             &ctx->salt_len))
-                return 0;
-        }
+        OPENSSL_free(ctx->salt);
+        ctx->salt = NULL;
+        if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->salt, 0,
+                                         &ctx->salt_len))
```

Comment:
Would this be more correct allowing for using NULL, 0 for the param?
```suggestion
        ctx->salt = NULL;
        ctx->salt_len = 0;
        if (p->data_size != 0 && p->data != NULL
            && !OSSL_PARAM_get_octet_string(p, (void **)&ctx->salt, 0,
                                            &ctx->salt_len))
```

Reply:
I thought about doing it this way.  The big issue for me is that we don't do it this way anywhere else.
I.e. we do not genreally:
- reset the length when setting a field to `NULL`
- check for `->data_size != 0`
- check for `->data == NULL`

If the param exists we just assume it has a valid pointer associated with it.

Reply:
I think @t8m is correct. Such a check is necessary because `OSSL_PARAM_get_octet_string()` fails if the parma's `data` is NULL. This currently actually prevents setting the salt to NULL.

Reply:
> I think @t8m is correct. Such a check is necessary because `OSSL_PARAM_get_octet_string()` fails if the param's `data` is NULL. This currently actually prevents setting the salt to NULL.

Not sure if this was an intended change, but we found this statement to be true. We can no longer set the salt to `NULL` in openssl 3.5.1. We get the following error:
```
80476A18AA750000:error:078C0102:common libcrypto routines:get_string_internal:passed a null parameter
```

Reply:
You could **never** set the salt to `NULL` before 3.5.1.  If you tried **the salt remained unchanged** and the call apparently suceeded.  So if you ever set the salt to a non–empty value, you could not set it back to empty<sup>†</sup>.  What changed in 3.5.1 is that trying to set the salt to `NULL` raises an error.  I think this is better than silently doing nothing and it's consistent with other parameter usage.

I see a few ways forward:
1. if, as is likely, you never set the salt explicitly, there is no need to reset it;
2. if you do set the salt explicitly, you're in a world of hurt because the previous salt will have been used without you knowing or
3. set the salt to empty by passing `""` and a length of zero instead of a NULL pointer.

In 3.4.0, line 310 means your attempt to set the parameter is silently ignored: https://github.com/openssl/openssl/blob/636dfadc70ce26f2473870570bfd9ec352806b1d/providers/implementations/kdfs/hkdf.c#L309-L317

In 3.5.1 this changed to: https://github.com/openssl/openssl/blob/f4fb4a65d5de45fa84665c30ecaf509ae27302a6/providers/implementations/kdfs/hkdf.c#L309-L315
The call on line 313 raises an error if the parameter's data pointer is NULL.

<sup>†</sup>: the reset call sets the salt to empty but also wipes everything else.